test/model/typing_status_test.dart

```diff
@@ -0,0 +1,207 @@
+    notifiedCount = 0;
+    model.addListener(() => notifiedCount += 1);
+  }
+
+  void checkNotNotified() {
+    check(notifiedCount).equals(0);
+  }
+
+  void checkNotifiedOnce() {
+    check(notifiedCount).equals(1);
+    notifiedCount = 0;
+  }
+
+  void checkTypists(Map<SendableNarrow, List<User>> typistsByNarrow) {
+    final actualTypistsByNarrow = <SendableNarrow, Iterable<int>>{};
+    for (final narrow in model.debugActiveNarrows) {
+      actualTypistsByNarrow[narrow] = model.typistIdsInNarrow(narrow);
+    }
+    check(actualTypistsByNarrow).deepEquals(
+      typistsByNarrow.map((k, v) => MapEntry(k, v.map((e) => e.userId))));
+  }
+
+  final stream = eg.stream();
+  final topicNarrow = TopicNarrow(stream.streamId, 'foo');
+
+  final dmNarrow = DmNarrow.withUser(eg.otherUser.userId, selfUserId: eg.selfUser.userId);
+  final groupNarrow = DmNarrow.withOtherUsers(
+    [eg.otherUser.userId, eg.thirdUser.userId], selfUserId: eg.selfUser.userId);
+
+  group('handle typing start events', () {
+    test('add typists in separate narrows', () {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser], groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('add a typist in the same narrow', () {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser, eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('sort dm recipients', () {
+      prepareModel();
+
+      final eventUnsorted = TypingEvent(id: 1, op: TypingOp.start, senderId: eg.otherUser.userId,
+        messageType: MessageType.direct,
+        recipientIds: [5, 4, 10, 8, 2, 1, eg.selfUser.userId],
+        streamId: null,
+        topic: null);
+      // DmNarrow's constructor expects the recipient IDs to be sorted.
+      check(() => model.handleTypingEvent(eventUnsorted)).returnsNormally();
+    });
+
+  });
+
+  group('handle typing stop events', () {
+    test('remove a typist from an unknown narrow', () async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.stop, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotNotified();
+    });
+
+    test('remove one of two typists in the same narrow', () async {
```

Comment:
Let's also have a test for removing the last typist in a narrow, checking that we don't leak an empty list for that narrow into our map.

… On rereading, I see you do cover that case at the end of "remove typists from different narrows". But it'd be good to have a separate test case just about that case, with a description to match. Otherwise it's easy to miss that that's there, as I did. Also if some future change were to accidentally break this behavior, if all that did was cause that last step of a larger test case to fail, it'd be easy to think it's just cleanup code and so miss that it's checking something meaningful, defeating the purpose of the test.

This is also one of the benefits of tightening the setup further, with something like `prepareModel(typists: …)` like in my immediate previous comment on the code just below here. When setup is concise, it's cheap to make more separate test cases and there isn't much reason to be tempted to combine conceptually different questions into a single test case.